File: Algorithms/mergesort.cpp
#include <bits/stdc++.h>

void merge(std ::vector<int> &v, int start, int mid, int end)
{
    std ::vector<int> temp(end - start + 1);
    int i = start, j = mid + 1, k = 0;
    while (i <= mid && j <= end)
    {
        if (v[i] <= v[j])
        {
            temp[k++] = v[i++];
        }
        else
        {
            temp[k++] = v[j++];
        }
    }
    while (i <= mid)
    {
        temp[k++] = v[i++];
    }
    while (j <= end)
    {
        temp[k++] = v[j++];
    }
    for (int i = start; i <= end; i++)
    {
        v[i] = temp[i - start];
    }
}

void mergesort(std ::vector<int> &v, int start, int end)
{
    if (start >= end)
        return;
    int mid = start + (end - start) / 2;
    mergesort(v, start, mid);
    mergesort(v, mid + 1, end);
    merge(v, start, mid, end);
}

int main()
{
    std ::vector<int> v{1, -1, 0, 9, -26, 76, 22, 73, 63};
    for (int i = 0; i < v.size(); i++)
    {
        std ::cout << v[i] << " ";
    }
    mergesort(v, 0, v.size() - 1);
    for (int i = 0; i < v.size(); i++)
    {
        std ::cout << v[i] << " ";
    }
    std ::cout << std ::endl;
}

File: Python/Algorithm/Merge-sort.py
def merge_sort(L):
    """Sort the array usin merge sorting"""

    # Calculate the value of middle index of array
    mid = int((len(L) - 1) / 2)

    # This loop will run when length of array is greater than 2
    # a is initialised with left side of the array
    # b is initialised with right side of the array
    if len(L) > mid + 2:
        a = merge_sort(L[0:mid])
        b = merge_sort(L[mid : len(L)])

    # This loop will when length of array is equal to 2
    # a is initiliased with the first element of array
    # b is initiliased with the second element of array
    elif len(L) == 2:
        a = [L[0]]
        b = [L[1]]

    else:
        return L

    i = 0
    j = 0
    new = []

    # This loop will run until
    # i is not equal to the lenth of array a OR
    # j is equal to length of array b
    while i != len(a) or j != len(b):

        # Checking if value of i and j is lesser than length of array a and b
        if i < len(a) and j < len(b):
            # If the element on the left side is lesser
            # The element on the right side
            # Left side element is added to the array new
            if a[i] < b[j]:
                new += [a[i]]
                i += 1

            # If element on the right side is lesser
            # The element on the left side
            # Right side element is added to the array new
            else:
                new += [b[j]]
                j += 1

        # If i gets equal to the length of array a
        # All the elements of array b are directly added to the array new
        elif i == len(a) and j != len(b):
            new += [b[j]]
            j += 1

        # If j gets equal to the length of array b
        # All the elements of array a are directly added to the array new
        elif j == len(b) and i != len(a):
            new += [a[i]]
            i += 1

        else:
            break

    return new


if __name__ == "__main__":
    array = list(map(int, input("Enter array::\t").split()))
    print(merge_sort(array))
